File: Uniform dispersion, 6.6.2017/dispersion.py
# -*- coding: utf-8 -*-
"""
Created on Sun Jun 04 10:06:23 2017

@author: chenpq
"""

from __future__ import division
import sys
import numpy as np
import math
from angle_between_two_vectors import angle 

#Don't generate .pyc file 
sys.dont_write_bytecode = True

# The function below is used execute the Dispersion Algorithm.
# To calculate the new position(coordinates) and velocity(speed/heading) for zebro.
# Input:
#   zebros:
#       zebros[i, :]: information of zebro_i
#       zebros[:, 1]: x coordinate of zebro
#       zebros[:, 2]: y coordinate of zebro
#       zebros[:, 3]: speed on x coordinate of zebro
#       zebros[:, 4]: speed on x coordinate of zebro
#       zebros[:, 5]: heading of zebro
#   speed: speed limitation on zebro
#   nNeighbors: Maximum number of neighbors to avoid
#   disDanger: Dangerous distance
#   disDisp: Neighbor in disDisp may have repusion on the zebro
# Output:
#   value: update of zebros. The data structure is the same with input 'zebros'. 
# Written by Pengqi Chen (chenpq1993@gmail.com).


def new_zebros(zebros, speed, nNeighbors, disDanger, disDisp):
    nNeighbors = int(nNeighbors)
    nZebros = len(zebros[:,0])
    # The code is for zebro 6. Python starts from 0.
    value = np.zeros((nZebros, 5))
                
    flag=0; # 0, Initial value
            # 1, Need to avoid collision with other zebros
    
    zebro = zebros[5] 
    v_x = zebro[2]
    v_y = zebro[3]
    v = 0
    r_angle_pi = zebro[4]
       
    # Calculate the distance from zebro i to all zebros(including itself)
    # and store the results in distance
    distance = np.zeros(nZebros)
    for j in range(0, nZebros):
        other = zebros[j]
        dis = math.sqrt(pow((zebro[0] - other[0]),2) +  pow((zebro[1] - other[1]),2))
        distance[j] = dis
                    
        
    # To sort the distance and put the sorted value in distance2(from lower to higher) 
    # To record the orignal number of sorted distance in order.
    distance2 = sorted(distance)
    order = sorted(range(len(distance)), key=lambda k: distance[k])
        

    count = 0
    # neighbor: [x, y, v_x, v_y, r_angle_pi, dist_x, dist_y, dist]
    neighbor = np.zeros((nNeighbors, 8))
    # 1, 2...num_Neighbros
    for n in range(0, nNeighbors):
        neighbor[n, 0:5] = zebros[order[n+1]]
        neighbor[n, 5] = zebro[0] - neighbor[n,0]
        neighbor[n, 6] = zebro[1] - neighbor[n,1]
        neighbor[n, 7] = math.sqrt(pow(neighbor[n,5], 2) + pow(neighbor[n,6], 2))
        if(distance2[n] < disDisp):
            count = count+1

            
    # There is only one neighbor in the range of disDanger
    con1 = distance2[1] < disDanger
    con2 = distance2[2] > disDanger
    if con1 and con2:
        flag = 1
        count = 1
   
    # To calculate the new velocity on x and y direction
    # k0: to keep the inertia of the movement of zebro 
    # k1: a coefficient to calculate the acceleration from distance
    a_x = 0
    a_y = 0
    a = 0
    a_ceil = speed / 3; # the ceiling bound of accelaration
    if flag == 0:
        k0 = 3.5/5
        k1 = 8
    elif flag == 1:
        k0 = 2/5
        k1 = 8
    
    if count == 0:    #No neighbor zebros is in the range of disDisp
        v_x = k0 * v_x
        v_y = k0 * v_y
            
    else:
        for p in range(0, count):
            dist_x = neighbor[p, 5]
            dist_y = neighbor[p, 6]
            dist3  = neighbor[p, 7]
            # To calculate accelaration
            if(dist3 < disDisp):  
                a_px = k1 / dist3 * dist_x / dist3
                a_py = k1 / dist3 * dist_y / dist3
                a_x = a_x + a_px
                a_y = a_y + a_py
             
            if(p == count-1):
                a = math.sqrt(pow(a_x, 2) + pow(a_y, 2))
                # To limit abs(a) to be in the range of speed/3
                a_x = a_x * (a_ceil) / max(a, a_ceil)
                a_y = a_y * (a_ceil) / max(a, a_ceil)
    
                # To calculate the new speed
                v_x = zebro[2] * k0 + a_x
                v_y = zebro[3] * k0 + a_y
                ##To limit abs(v) to be in the range of speed.
                v = math.sqrt(pow(v_x, 2) + pow(v_y, 2))
                v_x = v_x * speed / max(v, speed)
                v_y = v_y * speed / max(v, speed)
        
    # To eliminate jitter                     
    speed_floor = 0.2
    if(v < speed_floor):
        v_x = 0
        v_y = 0
            
        
    # To calculate the new heading 
    r_angle = np.zeros(2)
    if (v > speed_floor):
        # Calculate the heading of the zebro(direction of velocity)
        # with positive axis of yas the refrence, 
        # r_angle_pi>0 when x>0, r_angle_pi< 0 when x<0;
        # the range of r_angle_pi is (-pi, pi]
        m = [0, 1]
        n = [v_x, v_y]
        r_angle = angle(m, n)
        r_angle_pi = r_angle[0]
        if(v_x >= 0): 
            r_angle_pi = -r_angle_pi
    else:
        r_angle_pi = zebro[4]
                              
    if(r_angle_pi > math.pi):
        r_angle_pi = -r_angle_pi + 2*math.pi
    elif(r_angle_pi < -math.pi):
        r_angle_pi = 2*math.pi + r_angle_pi
        
    # output of the function
    value = [v_x + zebro[0], v_y + zebro[1], v_x, v_y, r_angle_pi]
    return value

File: Uniform Dispersion(Coverage), 6.19.1.2017/dispersion.m
function [zebroi] = dispersion(neiInRan, speed)
% Dispersion Algorithm 
% Input:
%   neiInRan: information of neighbor zebros in the range of zebroi.
%     neiInRan(:, 1): rel.distance from the neighbor zebro to zebroi 
%     neiInRan(:, 2): rel.angle between the neighbor zebro and zebroi
%     neiInRan(:, 3): Id of the neighbor zebro
%   speed: speed of zebroi
% Important Variable:
%   sCeil: speed limitation on zebro
%   aCeil: acceleration limitation on zebro
%   numNeighbors: maximum number of zebros to avoid
%   disDanger: dangerous distance
%   disDisp: neighbor in disDisp may have repusion on the zebro
% Output: 
%   zebroi: [v, tAngle]
%     v: new speed of zebroi
%     tAngle: turnning angle of zebroi
% Written by Pengqi Chen (chenpq1993@gmail.com).

% First check there is at least one close neighbor and than do the
% following calculation.
% if input_angle > 0, the neighbor is on the clockwise direction. 
% turn_angle > 0, turn right.
% if input_angle < 0, the neighbor is on the anticlockwise direction. 
% turn_angle < 0, turn left.
% the angle of acceleration uses the direction of the velocity as the basis.
%
% neiZebros = [1, pi/3; 1.333, -pi/3; 8, 0;];
numNeighbors = 3;                         % max number of neighbors to avoid.
disDisp = 10;
disDanger = 5;
distance = neiInRan(:, 1);        
[distance2, order] = sort(distance);
nneiZebros = 0;                             % number of neighbor zebros
numNeighbors = min(numNeighbors, length(distance));
neiZebros = zeros(numNeighbors, 3);   % neighbor zebros for calculation
for jzebro = 1:numNeighbors
    if(distance2(jzebro) < disDisp)
        nneiZebros = nneiZebros + 1;       
        neiZebros(nneiZebros,: ) = neiInRan(order(jzebro), :); % to get numNeighbors closest neighbor zebros
    end
end
% To calculate acceleration:
neiZebros = neiZebros(1: nneiZebros, :);
% To get the angle for acceleration:
for iIter = 1:nneiZebros
    neiZebros(iIter, 2) = acceAngle(neiZebros(iIter, 2));
end
aCeil = 1;
sCeil = 3;
ka = 5;
if nneiZebros == 0                        % No neighbor zebro in disDisp
    a = 0;
elseif nneiZebros == 1                    % One zebro in disDisp
    a = ka/neiZebros(1, 1);
    a = a * (aCeil) / max(a, aCeil);
    aAngle = neiZebros(1, 2);
else                                    % Equal to or More than 2 zebros in disDisp
    con1 = distance2(1) < disDanger;
    con2 = distance2(2) > disDanger;
    if con1 && con2
        ka = 8;
    end
    for jIter = 1:nneiZebros-1
        a1 = ka/neiZebros(1, 1);             % value of basis acceleration
        angle1 = neiZebros(1, 2);         % direction of basis acceleration
        a2 = ka/neiZebros(jIter + 1, 1);     
        angle2 = neiZebros(jIter + 1, 2);
        angle = angle1 - angle2;
        % angle is in range of (-pi, pi], a1 is the basis. 
        % angle > 0, a2 is on the anticlockwise direction
        % angle < 0, a2 is on the clockwise direction
        angle = limAn2Pi(angle);
        % value of the new basis acceleration
        a = sqrt(a1^2 + a2^2 + 2 * a1 * a2 * cos(angle)); 
        if a == 0               % if it is the last iteration, it also does not matter
            neiZebros(1, 1) = 0;                      
            neiZebros(1, 2) = 0;
            if(jIter == nneiZebros-1)
                aAngle = 0;
                break;
            end
            continue;
        elseif a1==0
            neiZebros(1, 1) = a2;
            neiZebros(1, 2) = angle2;
        else
            x = (a1^2 + a^2 - a2^2)/(2*a1*a);
            tangle = acos(x);
            if(angle < 0)           
                tangle = -tangle;
            end
            % The sum of acceleration is finished
            % Limit a to be in range of aCeil
            if(jIter == nneiZebros-1)
                a = a * (aCeil) / max(a, aCeil);
                aAngle = neiZebros(1, 2) - tangle;
                aAngle = limAn2Pi(aAngle);
            end
            % Next basis acceleration for calculation
            neiZebros(1, 1) = a;                      
            neiZebros(1, 2) = neiZebros(1, 2) - tangle;
            neiZebros(1, 2) = limAn2Pi(neiZebros(1, 2));
        end
    end 
end    

s = speed;
if a == 0 && s == 0
    s = 0;
    tAngle = 0;
elseif a == 0 && s ~= 0
    kv = 2/5;
    %kv = 0;
    s = kv * s;     %If there is no neighbor to avoid, just stop soon.
    tAngle = 0;
elseif a ~= 0 && s == 0
    s = a;
    tAngle = neiZebros(1, 2);
else
    kv = 3/5;
    %kv = 0;
    s = speed * kv;
    angle1 = aAngle;
    if angle1 == 0
        s = s + a;
        tAngle = 0;
    else
        s1 = sqrt(a^2 + s^2 + 2 * a * s * cos(angle1));
        if s1 == 0               % if it is the last iteration, it also does not matter
            s = 0;
            tAngle = 0;
        else
            x = (s^2 + s1^2 - a^2)/(2*s*s1);
            tAngle = acos(x);
            if angle1 < 0
                tAngle = -tAngle;           
            end
            s = s1;
        end
    end
end

sFloor = 0.2;   % Speed Floor
if s < sFloor
    s = 0;
    tAngle = 0; 
end

s = s * (sCeil) / max(s, sCeil);
tAngle = limAn2Pi(tAngle);
zebroi = [s, tAngle];